File: 2022/phase2_dress_rehearsal/byuvrx/src/frostlab-inekf-54148f8ac0c7/python/tests/Core/test_SO2.py
import numpy as np
from numpy.testing import assert_allclose
from scipy.linalg import expm, logm
from numpy.linalg import inv
from inekf import SO2
import pytest

def test_BaseConstructor1():
    mtx = np.eye(2)
    sigma = np.eye(3)
    A = np.ones(2)

    x = SO2[2](mtx, sigma, A)

    assert_allclose(mtx, x.State)
    assert_allclose(sigma, x.Cov)
    assert_allclose(A, x.Aug)

def test_BaseConstructor2():
    mtx = np.eye(2)
    sigma = np.eye(3)
    A = np.ones(2)

    with pytest.raises(Exception):
        SO2["D"](mtx, sigma)

    SO2["D"](mtx, sigma, A)
    SO2["D"]()

def test_ThetaConstructor():
    x = SO2(np.pi/4)
    v = 1 / np.sqrt(2)
    r = np.array([[v, -v],[v, v]])

    assert_allclose(r, x.State)

def test_TangentConstructor():
    x = np.arange(3)

    state = SO2["D"](x)

    assert_allclose(x[1:], state.Aug)
    assert_allclose(state.State, np.eye(2))

def test_AddAug():
    x = SO2["D"]()

    x.addAug(2)
    assert x.Aug[-1] == 2

    with pytest.raises(Exception):
        y = SO2()
        y.addAug(2)

def test_Inverse():
    x = SO2()
    assert_allclose(inv(x.State), x.inverse().State)

def test_Exp():
    x = np.arange(1,4)

    ours = SO2["D"].Exp(x)
    theirs = expm( SO2["D"].Wedge(x) )

    assert_allclose(theirs, ours.State)
    assert_allclose(x[-2:], ours.Aug)
    with pytest.raises(Exception):
        SO2[3].Exp(x)

def test_Log():
    xi = np.arange(3)
    x = SO2["D"](xi)

    assert_allclose(xi, x.log())

def test_Wedge():
    x = np.arange(1,4)
    ours = SO2["D"].Wedge(x)
    theirs =  np.array([[0, -1],
                        [1, 0]])
    
    assert_allclose(theirs, ours)

    with pytest.raises(Exception):
        SO2[3].Wedge(x)

def test_Adjoint():
    assert_allclose(np.ones(1), SO2().Ad())
    assert_allclose(np.eye(3), SO2[2]().Ad())

File: 2022/phase2_dress_rehearsal/byuvrx/src/frostlab-inekf-54148f8ac0c7/tests/Core/test_SO2.cpp
#include "gtest/gtest.h"
#include <Eigen/Core>
#include <unsupported/Eigen/MatrixFunctions>
#include <InEKF/Core>

#define EXPECT_MATRICES_EQ(M_actual, M_expected) \
  EXPECT_TRUE(M_actual.isApprox(M_expected, 1e-6)) << "  Actual:\n" << M_actual << "\nExpected:\n" << M_expected

#define PI    3.14159265358
#define SQRT2 1.41421356237

TEST(SO2, BaseConstructor1){
    Eigen::Matrix2d mtx = Eigen::Matrix2d::Identity();
    Eigen::Matrix3d sigma = Eigen::Matrix3d::Identity();
    Eigen::Vector2d A = Eigen::Vector2d::Ones();

    InEKF::SO2<2> x(mtx, sigma, A);

    EXPECT_MATRICES_EQ(mtx, x());
    EXPECT_MATRICES_EQ(sigma, x.Cov());
    EXPECT_MATRICES_EQ(A, x.Aug());
    EXPECT_TRUE(x.Uncertain());
}

TEST(SO2, BaseConstructor2){
    Eigen::Matrix2d mtx = Eigen::Matrix2d::Identity();
    Eigen::Matrix3d sigma = Eigen::Matrix3d::Identity();
    Eigen::VectorXd A = Eigen::Vector2d::Ones();

    InEKF::SO2<Eigen::Dynamic> x(mtx, sigma, A);

    // make sure it gets mad if we pass the wrong sigma
    EXPECT_THROW(InEKF::SO2<Eigen::Dynamic> x(mtx, sigma);, std::range_error);
    // Or doesn't if we pass the right one
    EXPECT_NO_THROW(  (InEKF::SO2<Eigen::Dynamic>(mtx, sigma, A)) );
    // default also works
    EXPECT_NO_THROW( (InEKF::SO2<Eigen::Dynamic>()) );
}

TEST(SO2, ThetaConstructor){
    InEKF::SO2<> x(PI/4);
    Eigen::Matrix2d r;
    r << 1/SQRT2, -1/SQRT2, 1/SQRT2, 1/SQRT2;

    EXPECT_MATRICES_EQ(x(), r);
}

TEST(SO2, TangentConstructor){
    Eigen::VectorXd x(3);
    x << 0, 1, 2;

    InEKF::SO2<Eigen::Dynamic> state(x);
    EXPECT_MATRICES_EQ(state.Aug(), x.tail(2));
    EXPECT_MATRICES_EQ(state(), Eigen::Matrix2d::Identity());
}

TEST(SO2, AddAug){
    InEKF::SO2<Eigen::Dynamic> x;
    EXPECT_EQ(x.Aug().rows(), 0);

    x.addAug(2);

    EXPECT_EQ(x.Aug()(0), 2);

    // TODO: Test adding to Cov

    InEKF::SE2<> y;
    EXPECT_THROW( y.addAug(2), std::range_error);
}

TEST(SO2, Inverse){
    InEKF::SO2<> x(PI/2);

    EXPECT_MATRICES_EQ(x.inverse()(), x().inverse());
}

TEST(SO2, Exp){
    Eigen::VectorXd x(3);
    x << 1,2,3;

    InEKF::SO2<Eigen::Dynamic> ours = InEKF::SO2<Eigen::Dynamic>::Exp(x);
    Eigen::Matrix2d theirs = InEKF::SO2<Eigen::Dynamic>::Wedge(x).exp();

    EXPECT_MATRICES_EQ(ours(), theirs);
    EXPECT_MATRICES_EQ(ours.Aug(), x.tail(2));
}

TEST(SO2, Log){
    Eigen::VectorXd x(3);
    x << 0, 1, 2;
    InEKF::SO2<Eigen::Dynamic> state(x);

    Eigen::Vector3d xi = state.log();
    EXPECT_MATRICES_EQ(xi, x);
}

TEST(SO2, Wedge){
    Eigen::VectorXd x(3);
    x << 1, 2, 3;

    Eigen::Matrix2d ours = InEKF::SO2<Eigen::Dynamic>::Wedge(x); 
    Eigen::Matrix2d theirs;
    theirs << 0, -1,
                1, 0;

    EXPECT_MATRICES_EQ(ours, theirs);
}

TEST(SO2, Adjoint){
    InEKF::SO2<1> x;
    EXPECT_MATRICES_EQ(x.Ad(), Eigen::Matrix2d::Identity());
    EXPECT_MATRICES_EQ(InEKF::SO2<1>::Ad(x), Eigen::Matrix2d::Identity());
}